lib/code_corps/github.ex

```diff
@@ -1,41 +1,272 @@
+    defp convert_error(%{"code" => code, "field" => field, "resource" => resource}) do
+      APIErrorObject.new([code: code, field: field, resource: resource])
+    end
+  end
+
+  defmodule HTTPClientError do
+    defstruct [:reason, message: """
+    The GitHub HTTP client encountered an error while communicating with
+    the GitHub API.
+    """]
+
+    def new(opts) do
+      struct(__MODULE__, opts)
+    end
+  end
+
+  @type method :: :get | :post | :put | :delete | :patch
+  @type headers :: %{String.t => String.t} | %{}
+  @type body :: {:multipart, list} | map
+  @typep http_success :: {:ok, integer, [{String.t, String.t}], String.t}
+  @typep http_failure :: {:error, term}
+
+  @type api_error_struct ::
+    %APIError{}            |
+    %HTTPClientError{}
+
+  @spec get_base_url() :: String.t
+  defp get_base_url() do
+    Application.get_env(:code_corps, :github_base_url) || "https://api.github.com/"
+  end
+
+  #
+  @spec get_token_url() :: String.t
+  defp get_token_url() do
+    Application.get_env(:code_corps, :github_oauth_url) || "https://github.com/login/oauth/access_token"
+  end
+
+  @spec use_pool?() :: boolean
+  defp use_pool?() do
+    Application.get_env(:code_corps, :github_api_use_connection_pool)
+  end
+
+  @spec add_default_headers(headers) :: headers
+  defp add_default_headers(existing_headers) do
+    Map.merge(%{"Accept" => "application/vnd.github.machine-man-preview+json"}, existing_headers)
+  end
 
-  Accepts a third parameter – a custom API module – for the purposes of
-  explicit dependency injection during testing.
+  @spec add_access_token_header(headers, String.t | nil) :: headers
+  defp add_access_token_header(existing_headers, nil), do: existing_headers
+  defp add_access_token_header(existing_headers, access_token) do
+    Map.put(existing_headers, "Authorization", "token #{access_token}")
+  end
 
-  Returns one of the following:
+  @doc """
+  Generates a JWT from the GitHub App's generated RSA private key using the
+  RS256 algo, where the issuer is the GitHub App's ID.
 
-  - `{:ok, %CodeCorps.User{}}`
-  - `{:error, %Ecto.Changeset{}}`
-  - `{:error, "some_github_error"}`
+  Used to exchange the JWT for an access token for a given integration, or
+  for the GitHub App itself.
+
+  Expires in 10 minutes.
   """
-  @spec connect(User.t, String.t, module) :: {:ok, User.t} | {:error, String.t}
-  def connect(%User{} = user, code, api \\ @api) do
-    case code |> api.connect do
-      {:ok, github_auth_token} -> user |> associate(%{github_auth_token: github_auth_token})
-      {:error, error} -> {:error, error}
+  def generate_jwt do
+    %{}
+    |> Joken.token
+    |> Joken.with_exp(Timex.now |> Timex.shift(minutes: 10) |> Timex.to_unix)
+    |> Joken.with_iss(@app_id |> String.to_integer())
+    |> Joken.with_iat(Timex.now |> Timex.to_unix)
+    |> Joken.with_signer(Joken.rs256(rsa_key()))
+    |> Joken.sign
+    |> Joken.get_compact
+  end
+
+  defp rsa_key do
+    JOSE.JWK.from_pem(@rsa_key)
```

Comment:
We store the contents. You'll see I added a `GITHUB_APP_PEM`, which is maybe a misnomer. This will be the contents of the RSA private key, and on Heroku will _also_ be stored as text. I believe Heroku allows up to 32kb of total data in your config vars.